xapian-core/diversify/diversify.cc

```diff
@@ -53,19 +53,22 @@ void
 Diversify::initialise_points(const MSet &source)
 {
     LOGCALL_VOID(API, "Diversify::initialise_points", source);
+    unsigned int count = 0;
     TermListGroup tlg(source);
     for (MSetIterator it = source.begin(); it != source.end(); ++it) {
-	points.push_back(Point(tlg, it.get_document()));
-	weights[*it] = it.get_weight();
+	points.emplace(*it, Xapian::Point(tlg, it.get_document()));
+	scores[*it] = it.get_weight();
+	// Initial top-k diversified documents
+	if (count < k) {
+	    main_dmset.push_back(*it);
+	    ++count;
+	}
     }
 }
 
 pair<Xapian::docid, Xapian::docid>
-Diversify::get_key(const Xapian::Point &doc_a, const Xapian::Point &doc_b)
+Diversify::get_key(const Xapian::docid &docid_a, const Xapian::docid &docid_b)
```

Comment:
Passing values of types where copying is expensive by `const` reference is good, but `Xapian::docid` is just an unsigned integer type for which copying is very cheap - likely cheaper than passing a pointer and dereferencing it how passing by `const` reference is probably implemented.

Reply:
This one is still pending.

Reply:
Oops, my bad. Modification pushed in a fixup commit. Thanks for pointing it out.